src/main/java/org/kiwiproject/consul/util/failover/strategy/RoundRobinConsulFailoverStrategy.java

```diff
@@ -0,0 +1,116 @@
+import static org.kiwiproject.consul.util.Lists.isNotNullOrEmpty;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.net.HostAndPort;
+import okhttp3.HttpUrl;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A {@link ConsulFailoverStrategy} that round-robins a list of Consul servers.
+ */
+public class RoundRobinConsulFailoverStrategy implements ConsulFailoverStrategy {
+
+    // set to invalid index for initial state, so that we always try the first target initially
+    @VisibleForTesting
+    final AtomicInteger lastTargetIndex = new AtomicInteger(-1);
+
+    private final List<HostAndPort> targets;
+    private final int numberOfTargets;
+    private final Duration delay;
+    private final boolean delayAfterFailedRequest;
+
+    /**
+     * Create a new instance with the given list of Consul target servers.
+     *
+     * @param targets the Consul servers
+     */
+    public RoundRobinConsulFailoverStrategy(List<HostAndPort> targets) {
+        this(targets, Duration.ZERO);
+    }
+
+    /**
+     * Create a new instance with the given list of Consul target servers.
+     * <p>
+     * After each failed request, the instance will sleep for the given delay.
+     *
+     * @param targets           the Consul servers
+     * @param delayAfterFailure the amount to sleep after a failed request
+     */
+    public RoundRobinConsulFailoverStrategy(List<HostAndPort> targets, Duration delayAfterFailure) {
+        checkArgument(isNotNullOrEmpty(targets), "targets must not be null or empty");
+        this.targets = List.copyOf(targets);
+        this.numberOfTargets = targets.size();
+
+        checkArgument(nonNull(delayAfterFailure), "delayAfterFailure must not be null");
+        var millis = delayAfterFailure.toMillis();
+        checkArgument(millis >= 0, "delayAfterFailure must be zero or a positive duration");
+        this.delay = delayAfterFailure;
+        this.delayAfterFailedRequest = millis > 0;
+    }
+
+    @Override
+    @NonNull
+    public Optional<Request> computeNextStage(Request previousRequest) {
+        return computeNextStage(previousRequest, null);
+    }
+
+    @SuppressWarnings("removal")
+    @Override
+    @NonNull
+    public Optional<Request> computeNextStage(@NonNull Request previousRequest, @Nullable Response previousResponse) {
+        var nextTarget = hostAndPortFromOkHttpRequest(previousRequest);
+        var indexOfNextTarget = targets.indexOf(nextTarget);
+
+        if (lastTargetIndex.get() == indexOfNextTarget) {
+            var nextIndex = nextCircularListIndex(indexOfNextTarget);
+
+            lastTargetIndex.set(nextIndex);
```

Comment:
this is suspect to race condition since the `get` and the `set` are not synchronized.  you can try `compareAndSet()` to test out the old value and the new value to see if was set.

for example, let say `lastTargetindex = 0`, and `indexOfNextTarget = 0`
1. t1 runs and the condition matches, so it goes into this block, and yields
2. t2 runs and the condition matches, so it goes into this block, and yields
3. t1 runs to where it sets the nextIndex, and yields
4. t2 runs to the setting of the nextIndex (compareAndSet will tell if it was set)

Reply:
Good catch, makes sense to change line 81 to:

```java
lastTargetIndex.compareAndSet(indexOfNextTarget, nextIndex);
```

That's what you were saying, correct?

Reply:
Yeah, totally makes sense.